### PR TITLE
feat: include object id/key in invalid object errors

### PR DESCRIPTION
**Pull request by stoneman**
## Summary
- Include object `id`/`key` in errors when objects not found in datafile
- Modifies invalid object `id`/`key` log messages to make them consistent

Include object `id`/`key` in errors when an object is not found makes them available to the user or the custom error handler. 

One example of why this is useful is that the `key` of an experiment could be used within a custom error handler to fetch the details of the experiment. This would indicate whether the experiment has been paused (in which case the error could be ignored) or archived (in which case the code referencing the experiment should be removed from the application).

This change will impact existing custom error handlers which are expecting an error class rather than an error object instance.

## Test plan

Jira ticket:
https://jira.sso.episerver.net/browse/FSSDK-9439

### CODE REVIEW

**Comment by mikechu-optimizely** [comment]
We're super late on looking at this @stoneman.  Sorry about that. 

Is this change still relevant?

File: lib/optimizely/exceptions.rb [hunk 1]
```diff
@@ -74,24 +86,56 @@ def initialize(msg = 'Event tags provided are in an invalid format.')
...
-      super
+    attr_reader :experiment_id, :experiment_key
+
+    def initialize(experiment_id: nil, experiment_key: nil)
+      raise ArgumentError, 'Either experiment_id or experiment_key must be provided.' if experiment_id.nil? && experiment_key.nil?
+      raise ArgumentError, 'Cannot provide both experiment_id and experiment_key.' if !experiment_id.nil? && !experiment_key.nil?
+
+      if experiment_id.nil?
+        @experiment_key = experiment_key
+        indentifier = "key '#{@experiment_key}'"
```
Comment (Mat001) [inline lib/optimizely/exceptions.rb:97]: a typo?

File: lib/optimizely/exceptions.rb [hunk 2]
```diff
@@ -74,24 +86,56 @@ def initialize(msg = 'Event tags provided are in an invalid format.')
...
+      else
+        @experiment_id = experiment_id
+        indentifier = "id '#{@experiment_id}'"
```
Comment (Mat001) [inline lib/optimizely/exceptions.rb:100]: typo --> identifier?

File: lib/optimizely/exceptions.rb (hunk 1 above)
Comment (Mat001) [inline lib/optimizely/exceptions.rb:97]:
```suggestion
        identifier = "key '#{@experiment_key}'"
```

File: lib/optimizely/exceptions.rb (hunk 2 above)
Comment (Mat001) [inline lib/optimizely/exceptions.rb:100]:
```suggestion
        identifier = "id '#{@experiment_id}'"
```

File: lib/optimizely/exceptions.rb [hunk 3]
```diff
@@ -74,24 +86,56 @@ def initialize(msg = 'Event tags provided are in an invalid format.')
...
+      end
+
+      super("Experiment #{indentifier} is not in datafile.")
```
Comment (Mat001) [inline lib/optimizely/exceptions.rb:103]:
```suggestion
      super("Experiment #{identifier} is not in datafile.")
```

File: lib/optimizely/exceptions.rb [hunk 4]
```diff
@@ -74,24 +86,56 @@ def initialize(msg = 'Event tags provided are in an invalid format.')
...
-    def initialize(msg = 'Provided variation is not in datafile.')
-      super
+    attr_reader :variation_id, :variation_key
+
+    def initialize(variation_id: nil, variation_key: nil)
+      raise ArgumentError, 'Either variation_id or variation_key must be provided.' if variation_id.nil? && variation_key.nil?
+      raise ArgumentError, 'Cannot provide both variation_id and variation_key.' if !variation_id.nil? && !variation_key.nil?
+
+      if variation_id.nil?
+        indentifier = "key '#{variation_key}'"
```
Comment (Mat001) [inline lib/optimizely/exceptions.rb:131]:
```suggestion
        identifier = "key '#{variation_key}'"
```

File: lib/optimizely/exceptions.rb [hunk 5]
```diff
@@ -74,24 +86,56 @@ def initialize(msg = 'Event tags provided are in an invalid format.')
...
+        @variation_key = variation_key
+      else
+        indentifier = "id '#{variation_id}'"
```
Comment (Mat001) [inline lib/optimizely/exceptions.rb:134]:
```suggestion
        identifier = "id '#{variation_id}'"
```

File: lib/optimizely/exceptions.rb [hunk 6]
```diff
@@ -74,24 +86,56 @@ def initialize(msg = 'Event tags provided are in an invalid format.')
...
+        @variation_id = variation_id
+      end
+
+      super("Variation #{indentifier} is not in datafile.")
```
Comment (Mat001) [inline lib/optimizely/exceptions.rb:138]:
```suggestion
      super("Variation #{identifier} is not in datafile.")
```

**Review by Mat001 — APPROVED** [review]

**Comment by stoneman** [comment]
@Mat001 signed! :)

**Pull request merged**